### PR TITLE
ShaderChunk to ES6 Module

### LINKED ISSUE #14446: Three.js source does not work as ES6 module
Ideally, developers should be able to use Three.js in pure ES6 module form.

For example, the following should work:

```HTML
<script type="module">
import * as THREE from "./src/Three.js";
</script>
```

Instead, we get errors like this one:

```
Failed to load module script: The server responded with a non-JavaScript MIME type of "application/octet-stream". Strict MIME type checking is enforced for module scripts per HTML spec.
``` 

This happens because glsl shader chunks are not real modules and need to be processed during build step by rollup glsl plugin.

Fixing this would improve development experience and also help with #9403 and #9562

Suggested solutions:

1. Rollup ShadersChunks.js into ShaderChunksBuild.js separately before building the library.

2. Convert each individual glsl shader chunk into ES6 module (automatically).

3. Rewrite and maintain shader chunks as ES6 modules using template literals. For example:

```javascript
var encodings_fragment = glsl`
 gl_FragColor = linearToOutputTexel( gl_FragColor );
`;
export {encodings_fragment};
```

##### Three.js version

- [x ] Dev

##### Browser

- [x] All of them

##### OS

- [x] All of them

### PR DESCRIPTION
**Pull request by arodic**
Added another step to rollup.config.js to make ShaderChunk.js compatible with ES6 modules.

Fixes #14446

### CODE REVIEW

**Comment by mrdoob** [comment]
Very nice!

File: src/renderers/shaders/ShaderChunk.js [hunk 1]
```diff
@@ -1,121 +1,234 @@
...
-import sprite_vert from './ShaderLib/sprite_vert.glsl';
-
-export var ShaderChunk = {
+var alphamap_fragment = "#ifdef USE_ALPHAMAP\r\n\r\n\tdiffuseColor.a *= texture2D( alphaMap, vUv ).g;\r\n\r\n#endif\r\n";
+
+var alphamap_pars_fragment = "#ifdef USE_ALPHAMAP\r\n\r\n\tuniform sampler2D alphaMap;\r\n\r\n#endif\r\n";
+
+var alphatest_fragment = "#ifdef ALPHATEST\r\n\r\n\tif ( diffuseColor.a < ALPHATEST ) discard;\r\n\r\n#endif\r\n";
+
+var aomap_fragment = "#ifdef USE_AOMAP\r\n\r\n\t// reads channel R, compatible with a combined OcclusionRoughnessMetallic (RGB) texture\r\n\tfloat ambientOcclusion = ( texture2D( aoMap, vUv2 ).r - 1.0 ) * aoMapIntensity + 1.0;\r\n\r\n\treflectedLight.indirectDiffuse *= ambientOcclusion;\r\n\r\n\t#if defined( USE_ENVMAP ) && defined( PHYSICAL )\r\n\r\n\t\tfloat dotNV = saturate( dot( geometry.normal, geometry.viewDir ) );\r\n\r\n\t\treflectedLight.indirectSpecular *= computeSpecularOcclusion( dotNV, ambientOcclusion, material.specularRoughness );\r\n\r\n\t#endif\r\n\r\n#endif\r\n";
```
Comment (mrdoob) [inline src/renderers/shaders/ShaderChunk.js]: Seems like the glsl() plugin is not cleaning up the code comments anymore?

File: rollup.config.js [hunk 2]
```diff
@@ -23,23 +23,35 @@ function glsl() {
...
+export default [
+	{
+		input: 'src/renderers/shaders/ShaderChunkSrc.js',
+		plugins: [
+			glsl()
+		],
+		output: [
+			{
+				format: 'es',
+				file: 'src/renderers/shaders/ShaderChunk.js',
```
Comment (Mugen87) [inline rollup.config.js]: So a build-result is now part of `src` directory?

**Comment by arodic** [comment]
> So a build-result is now part of the src directory?

Sort of... It's a compiled ES6 version of glsl files which otherwise cannot be used as modules. Alternatively, the file can sit outside src/ however, modules that need it should have import paths pointing to built version, not .glsl. I'm not sure I like `import { ShaderChunk } from "../../../build/ShaderChunk.js"` better.

**Comment by arodic** [comment]
@looeee I'm not sure what you mean by that? As far as I can tell, making any changes to glsl files requires a build step. This PR only splits the build I'm two steps to unblock ShaderChunk from breaking otherwise working modular architecture.

**Comment by Mugen87** [comment]
Artifacts of a build should definitely not go into the source directory. If possible, use the `build` directory.

**Comment by looeee** [comment]
> As long as you continue building after changing glsl (same as before) nothing should be out of sync

This is what I mean - if I submit a PR to update or add a glsl file, once that PR is accepted the src/ files will be out of sync in the dev branch, unless I also run build and include the build files in the PR.

Since build currently builds all the files, we would need at least a separate npm task for this. 

Conceptually, it's a little odd to have files from a build process in src/, although an argument can be made as to why this should be an exception, since the user will be using the built files in a secondary build process.

But then again, it's not hard for the user to include the rollup glsl() plugin. Unless there are other build systems for which this is an issue, perhaps we could just document how to do that instead and avoid adding this entirely?

**Comment by arodic** [comment]
> [Perhaps] this should be an exception since the user will be using the built files in a secondary build process.
> But then again, it's not hard for the user to include the rollup glsl() plugin.

My hope is to make threejs work without any build process whatsoever since modules and http2 has landed in most browsers. Of course - bundling, minifying, tree shaking etc. will be necessary for most projects. But increasing number of projects could benefit from using threejs in its source form.

So from that perspective, glsl files converted to js are still javascript source files and belong to src/ from javascript perspective. Unfortunately, since there is a "build" step to convert glsl to js, ShaderChunk.js in not a true source file.

Perhaps it would be less confusing if glsl files are converted to js individually (not bundled)?

Ideally, I'd like to have shader files consumable without any "build" process. As I already mentioned, there are alternative solutions such as to write shaders source in template literals (which are quite powerful and can remove comments btw).

Another option is to use XHR with http2-push but I believe it requires a custom server...

Anyway... I am happy with this as a temporary solution for my project. I can amend this PR if there is consensus on how to resolve the brought up issues. Otherwise, I'll be closing this PR.

**Comment by cecilemuller** [comment]
If we want to avoid a build step, either the glsl files have be converted to js (and the glsl files removed), either load glsl files at runtime with ajax (like one would load textures or other arbitrary assets).

I have a small preference for the second solution, but either way would be fine imho.

**Comment by looeee** [comment]
> My hope is to make threejs work without any build process whatsoever since modules and http2 have landed in most browsers.

Hmm, that is a good point. On the other hand, we may possibly never reach the time where importing thousands of small files is as performant as importing ten or so larger ones, since you will still want to do minification and compression. Compression will not work as well on small files as it does on large ones, while minification will at least still require a build step of some kind. 

There is also an overhead associated with importing ES6 modules. If you are importing ten or twenty modules, you can probably ignore this. But if you are importing 1000 critical modules, it's going to add some significant time to your page load. 

> Perhaps it would be less confusing if glsl files are converted to js individually (not bundled)?

This would be less confusing, but for the reasons above is unlikely to ever be useful. 

> bundling, minifying, tree shaking etc. will be necessary for most projects. But increasing number of projects could benefit from using threejs in its source form.

Could you give an example of such a project? Since browsers being able to import modules is very new and I'm not fully familiar with it yet, perhaps I'm just not seeing the potential applications.

File: src/renderers/shaders/ShaderChunk.js [hunk 3]
```diff
@@ -1,121 +1,234 @@
...
-import normal_vert from './ShaderLib/normal_vert.glsl';
-import points_frag from './ShaderLib/points_frag.glsl';
-import points_vert from './ShaderLib/points_vert.glsl';
-import shadow_frag from './ShaderLib/shadow_frag.glsl';
-import shadow_vert from './ShaderLib/shadow_vert.glsl';
-import sprite_frag from './ShaderLib/sprite_frag.glsl';
-import sprite_vert from './ShaderLib/sprite_vert.glsl';
-
-export var ShaderChunk = {
+var alphamap_fragment = "#ifdef USE_ALPHAMAP\r\n\r\n\tdiffuseColor.a *= texture2D( alphaMap, vUv ).g;\r\n\r\n#endif\r\n";
```
Comment (pailhead) [inline src/renderers/shaders/ShaderChunk.js]: what exactly is this file now?

File: rollup.config.js (hunk 2 above)
Comment (pailhead) [reply rollup.config.js]: Yeah i'm confused, this file gets built, and then is used for another build?

**Comment by pailhead** [comment]
I've started exploring [how to remove the `ShaderChunk`](https://github.com/pailhead/three-refactor-chunk-material) stuff completely out of three.js. The idea is pretty simple:

1. assume that this works:

```
const myVertexShader = require('my.vert')
const myFragmentShader = require('my.frag')
const myShader = new THREE.ShaderMaterial({
  vertexShader: myVertexShader,
  fragmentShader: myFragmentShader,
  uniforms:{ foo: {value:0}, bar: {value: new THREE.Vector2()}
})
```

2. this should work too

```
const myVertexShader = parseIncludes( someTemplate.vert )
const myFragmentShader = parseIncludes( someTemplate.frag)
const myShader = new THREE.ShaderMaterial({
  vertexShader: myVertexShader,
  fragmentShader: myFragmentShader,
  uniforms:{ 
    diffuse: {value: new THREE.Color},
    diffuseMap: {value: null }
  }
})
```

I'm under the impression, that for my needs, this introduces a lot of flexibility, and i can nuke a lot of code from `THREE.WebGLRenderer`. However, i've encountered concerns about tree shaking a few times, but it's really hard to gather all the bits and pieces and i have to admit that i don't know how it applies here.

Given that this is the most recent `THREE.ShaderChunk` discussion, could someone possibly elaborate more on the intricacies of tree shaking in this particular case. What is the beginning state, and what is the state after building or after importing stuff in some bundler?

**Comment by looeee** [comment]
@pailhead this PR is about importing GLSL files as ES6 modules in the browser (or in module bundlers). No matter what system we use for materials - the current system, your system, node materials etc. we will still have GLSL code that needs to be consumed by tools that are designed for JavaScript modules - this means that discussion of alternate approaches to materials, while interesting, is not relevant here. 

To summarise: 

* we have files that end in `.glsl` and we want to keep these for easy syntax highlighting. 
* we have build tools and importers that expect only valid `.js` files and we need a way to use the `.glsl` files in these. 

Possible solutions: 

* this PR: a pre-build step that imports all the `.glsl` files via `shaderChunkSrc.js` and bundles them into `shaderChunk.js`, which can then be consumed by browsers and bundlers. 
* a pre-build step that bundles each file seperately so that e.g.  `shadowmap_pars_vertex.glsl` becomes `shadowmap_pars_vertex.js`
* rewrite the `.glsl` files as `.js` using template literals and expect people to set up complex syntax highighting themselves (if that is even possible)
* do nothing (or write better documentation): it will not be possible to import the three.js src directly in the browser, and people bundling the source themselves have to add plugins such as `glslify` to their build step (as they currently do)

**Comment by pailhead** [comment]
>we have build tools and importers that expect only valid .js files and we need a way to use the .glsl files in these.

I don't understand this, could you please elaborate? 

If `we` refers to `people who write javascript` then your statement is wrong, and there are ways to transform various `.ext` files to something that javascript can use. 

Both your a) and b) mean this: (?)

```
npm run build:glsl
npm run build
```

>JavaScript modules - this means that discussion of alternate approaches to materials, while interesting, is not relevant here.

I think you misread something (you often seem to do so with my posts). You wrote:

> > bundling, minifying, tree shaking etc. will be necessary for most projects. But increasing number of projects could benefit from using threejs in its source form.
>
> Could you give an example of such a project? Since browsers being able to import modules is very new and I'm not fully familiar with it yet, perhaps I'm just not seeing the potential applications.

These two comments, from yourself and @arodic seem to tackle the material systems or at least the intricacies of building them (along with the rest of three). Not just how browsers import some js files in 2018. Feel free to clarify what you want discussed in this thread, but i'm curious about the same thing you are.

**Comment by looeee** [comment]
> we have build tools and importers that expect only valid .js files and we need a way to use the .glsl files in these.

> I don't understand this, could you please elaborate?

Sure, I should have said "bundlers" rather than build tools. Rollup.js is an example - you cannot pass a `.glsl` file into rollup without a plugin.  By importers, I mainly mean browsers, since you can now import ES6 modules directly in the browser.

**Comment by arodic** [comment]
> I don't think i'll ever end up in a situation where some .html page would load a bunch of atomic es6 modules from three.js

Here is where my motivation for this PR comes from... It's mostly about development process, not deployment.

I currently find myself in a situation where I experiment with UI framework designed around three.js while simultaneously I thinker with three.js source and examples/ From my process perspective, these components share the same source tree. This enables me to quickly explore ideas that require changes in three.js source, examples, and UI framework without any constraints or limitations.

At the same time, In my development environment, I have no issues loading thousands of files via module imports in a split of a second. Full threejs src reload with multiple example reload is so fast it adds pretty much zero overhead to my process thus I have no need for any build tools whatsoever #nonpminstall

Anyway,

Another thing to consider (future tech) is HTML modules. It should give as an option to include HTML files from js modules. If glsl code lives inside HTML modules with `<script type=glsl>` it could be the most graceful and #usetheplatform  solution for the above.

Thanks @looeee for the summary. Should I close this PR @mrdoob?

**Comment by tschoartschi** [comment]
> > @arodic 
> > My hope is to make threejs work without any build process whatsoever since modules and http2 have landed in most browsers.

> @looeee 
> Hmm, that is a good point. On the other hand, we may possibly never reach the time where importing thousands of small files is as performant as importing ten or so larger ones, since you will still want to do minification and compression. Compression will not work as well on small files as it does on large ones, while minification will at least still require a build step of some kind.

I think loading the shaders async could be an interesting approach. It would enalbe you to use the `glsl` files as they are (maybe do some minification, whitespace and comment removal, variable renaming for production). Furthermore it would trim down the JS size of Three.js which is important for many reasons, it's not only download time also JavaScript parsing can be expensive on some devices (very interesting comparision between [parsing JS and an JPEG](https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/javascript-startup-optimization/images/1_PRVzNizF9jQ_QADF5lQHpA.png), [full article here](https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/javascript-startup-optimization)). If you have more chunks it enables you to do better caching. If you would split the code into one JS file and one glsl file you could cache them seperately. If there are only changes in the JS file you only need to load the JS file and keep the glsl file cached and the other way around.

The Ember.js team are investigating something interesting for the templates (HTML files) in their framework. They split the HTML "code" from the JS code and serve the HTML "code" as binary data. This makes parsing very fast and helped them improve startup time a lot. More info [here](https://www.emberjs.com/blog/2017/10/10/glimmer-progress-report.html#toc_binary-templates), and a very interesting video [here](https://youtu.be/nXCSloXZ-wc?t=5m4s)

 I think something similar could be possible with glsl files as well.

But if you load glsl files async this makes everything which relies on these files async as well and this can get complicated very quickly. I'm not sure if this is worth the whole effort but definitly an interesting idea.

**Comment by pailhead** [comment]
I think every single user involved in this conversation has a different use case. @arodic doesn’t seem to use npm at all.

What you’re describing is totally possible, right here right now. You just have to write your own work around and three.js will still load the glsl through the built chunks thus doing it twice.

Removing the chunks from the core may be the common denominator that all of us people, and all of the proposals share.

**Comment by arodic** [comment]
Ok, since there is no clear path forward for this PR, I'll close it and update the issue with relevant information from this discussion.

**Pull request closed without merging**